src/Control/Monad/Logic/Sequence/Internal.hs

```diff
@@ -126,16 +128,44 @@ type Seq = SeqT Identity
 
 fromView :: m (View m a) -> SeqT m a
 fromView = SeqT . singleton
-{-# INLINE fromView #-}
+{-# INLINE[0] fromView #-}
 
 toView :: Monad m => SeqT m a -> m (View m a)
-toView (SeqT s) = case viewl s of
-  EmptyL -> return Empty
-  h S.:< t -> h >>= \x -> case x of
-    Empty -> toView (SeqT t)
-    hi :< SeqT ti -> return (hi :< SeqT (ti S.>< t))
-{-# INLINEABLE toView #-}
-{-# SPECIALIZE INLINE toView :: Seq a -> Identity (View Identity a) #-}
+toView (SeqT s0) = go s0 where
+  go s = case viewl s of
+    EmptyL -> return Empty
+    h S.:< t -> h >>= \x -> case x of
+      Empty -> go t
+      hi :< SeqT ti -> return (hi :< SeqT (ti S.>< t))
+{-# INLINE[0] toView #-}
+
+data Step s a = Done | Yield a s | Skip s
+data StreamM m a = forall s. StreamM (s -> m (Step s a)) s
+
+stream :: Monad m => SeqT m a -> StreamM m a
+stream m = StreamM next m where
+  {-# INLINE next #-}
+  next s = do
+    x <- toView s
```

Comment:
My kingdom for an `<$>`....

Reply:
I think it's better this way. `fmap` would introduce laziness I'm guessing we don't want.